File: python/app/samplegrid.py
from Vector3 import Vector3
from Vector2 import Vector2
from app.sample import Sample
import numpy as np
import sklearn.preprocessing
from app.vector import Vector
# ols
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

from scipy.optimize import leastsq

import time
class SampleGrid:

    def __init__(self,config):
        self.samples = []
        self.prunedSamples = []
        self.planes = []
        self.planeAvg = []
        self.floorSegment = []
        self.floorPlane = Sample()
        self.planeCluster = []
        self.upVector = Vector(0,0,0)
        self.upVector[2] = 1
        self.config = config
        self.floorPlane.n = self.upVector


    def initialize(self):
        self.samples.clear()
        for k in range(0, self.config["samplesY"]):
            V = []
            for l in range(0, self.config["samplesX"]):
                i = l * (self.config["IMAGE_WIDTH"]-1) / (self.config["samplesX"]-1)
                j = self.config["IMAGE_HEIGHT"]-1-k * (self.config["IMAGE_HEIGHT"]-1) / (self.config["samplesY"]-1)
                sample = Sample()
                sample.gridIdx = Vector(l, k)
                sample.imagePx = Vector(i, j)
                sample.bufferIdx = int(i+j * self.config["IMAGE_WIDTH"])
                V.append(sample)

            self.samples.append(V)





    def update(self,points):
        block_sum = 0
        start = time.time()
        if self.upVector* self.floorPlane.n > 0.5:
            self.setUpVector(self.floorPlane.n)
        else:
            pass

        for i in range(0, len(self.samples)):
            for j in range(0, len(self.samples[i])):
                self.samples[i][j].p = Vector(*(points[self.samples[i][j].bufferIdx].tolist()))
                self.samples[i][j].inGrid = self.samples[i][j].p.any()

        #normal = np.zeros((3,1))
        end = time.time()
        print("Part 1 took: " + str((end - start) * 1000) + "ms")
        start = time.time()
        for i in range(0, len(self.samples)):
            for j in range(0, len(self.samples[i])):
                if not self.samples[i][j].inGrid:
                    continue
                upIdx = i + 1
                if i == len(self.samples) - 1 or not self.samples[upIdx][j].inGrid:
                    upIdx = i

                downIdx = i - 1
                if i == 0 or not self.samples[downIdx][j].inGrid:
                    downIdx = i

                rightIdx = j + 1
                if j == len(self.samples[i]) - 1 or not self.samples[i][rightIdx].inGrid:
                    rightIdx = j

                leftIdx = j - 1
                if j == 0 or not self.samples[i][leftIdx].inGrid:
                    leftIdx = j

                if upIdx == downIdx or leftIdx == rightIdx:
                    continue

                up = self.samples[upIdx][j]
                down = self.samples[downIdx][j]
                right = self.samples[i][rightIdx]
                left = self.samples[i][leftIdx]

                start = time.time()
                normal = ((up.p - down.p) ^ (right.p - left.p))
                normal = -normal
                normal = normal.normalize()
                self.samples[i][j].n = normal
                end =time.time()
                block_sum += (end-start)
        end = time.time()
        print("Critical block took: " + str(block_sum * 1000) + "ms")

    def setUpVector(self,up):
        self.upVector = up
        self.upVector = self.upVector.normalize()
        Sample.up = self.upVector


    def findFloor(self):
        start = time.time()
        self.prune()
        if len(self.prunedSamples) < 2:
            return self.floorPlane

        #sort by z coordinate of point
        self.prunedSamples.sort(key=lambda x: x.p[2])#*self.upVector,reverse=True)
        #self.prunedSamples.sort()

        self.planes.clear()
        self.planeAvg.clear()
        self.floorSegment.clear()
        self.floorPlane = self.prunedSamples[1]
        end = time.time()

        print("Pruning: " + str((end - start) * 1000) + "ms")
        start = time.time()
        for i in range(2, len(self.prunedSamples) - 1):
            if not self.isIn(self.prunedSamples[i].gridIdx):
                continue


            self.planeCluster.clear()
            self.floodFill(self.prunedSamples[i].gridIdx)
            if len(self.planeCluster) == 1:
                continue
            avg = Sample()
            avg.n[2] = 0
            for  j in range(0, len(self.planeCluster)):
                avg += self.planeCluster[j]

            avg /= float(len(self.planeCluster))

            self.planeAvg.append(avg)
            self.planes.append(self.planeCluster)

            if self.floorPlane.distance(avg) < self.config["mergeThreshold"]:
                self.floorPlane.p = (self.floorPlane.p * len(self.floorSegment) + avg.p * len(self.planeCluster)) / (len(self.floorSegment) + len(self.planeCluster))
                self.floorPlane.n = (self.floorPlane.n * len(self.floorSegment) + avg.n * len(self.planeCluster)) / (len(self.floorSegment) + len(self.planeCluster))
                self.floorPlane.n = self.floorPlane.n.normalize()
                self.floorSegment.extend(self.planeCluster)

            if len(self.floorSegment) * 20 < len(self.planeCluster):
                self.floorPlane = avg
                self.floorSegment.clear()
                self.floorSegment.extend(self.planeCluster)
        end = time.time()
        print("floodfill: " + str((end - start) * 1000) + "ms")
        if len(self.floorSegment) > 2:
            start = time.time()
            data_x = np.array([np.array([x.p[0], x.p[1],x.p[2]]) for x in self.floorSegment])
            data_y = np.array([x.p[2] for x in self.floorSegment])
            solution = np.linalg.lstsq(data_x,data_y, rcond=None)[0]
            self.floorPlane.p[2] = self.floorPlane.p[0] * solution[0] + self.floorPlane.p[1] * solution[1] + solution[2]
            self.floorPlane.n = Vector(-solution[0], -solution[1], 1).normalize()
            end = time.time()
            print("leastsq: " + str((end - start) * 1000) + "ms")

            # start = time.time()
            # data = np.array([x.p for x in self.floorSegment]).transpose()
            # p, n = self.planeFit(data)
            # end = time.time()
            # print("SVD: " + str((end - start) * 1000) + "ms")
            # self.floorPlane.p = Vector(*(p.tolist()))
            # self.floorPlane.n = Vector(*(n.tolist())).normalize()



        return self.floorPlane




    def floodFill(self, parentIdx):
        parent = self.samples[parentIdx[1]][parentIdx[0]]
        if not parent.inGrid:
            return

        parent.inGrid = False
        self.planeCluster.append(parent)
        if parent.gridIdx[0] > 0:
            childIdx = parent.gridIdx - Vector(1, 0)
            child = self.samples[childIdx[1]][childIdx[0]]

            if parent.distance(child) < self.config["floodThreshold"]:
                self.floodFill(childIdx)

        if parent.gridIdx[0] < self.config["samplesX"] - 1:
            childIdx = parent.gridIdx + Vector(1, 0)
            child = self.samples[childIdx[1]][childIdx[0]]
            if parent.distance(child) < self.config["floodThreshold"]:
                self.floodFill(childIdx)

        if parent.gridIdx[1] > 0:
            childIdx = parent.gridIdx - Vector(0, 1)
            child = self.samples[childIdx[1]][childIdx[0]]
            if parent.distance(child) < self.config["floodThreshold"]:
                self.floodFill(childIdx)

        if parent.gridIdx[1] < self.config["samplesY"] - 1:
            childIdx = parent.gridIdx + Vector(0, 1)
            child = self.samples[childIdx[1]][childIdx[0]]
            if parent.distance(child) < self.config["floodThreshold"]:
                self.floodFill(childIdx)



    def prune(self):
        self.prunedSamples.clear()
        for i in range(0, len(self.samples)):
            for j in range(0, len(self.samples[i])):
                if not self.samples[i][j].inGrid:
                    continue

                self.samples[i][j].angle = self.samples[i][j].n * self.upVector


                if self.samples[i][j].angle > self.config["pruneThreshold"]:
                    self.prunedSamples.append(self.samples[i][j])
                else:
                    self.samples[i][j].inGrid = False


    def isIn(self, gridIdx):
        return self.samples[gridIdx[1]][gridIdx[0]].inGrid

    def planeFit(self,points):
        """
        p, n = planeFit(points)

        Given an array, points, of shape (d,...)
        representing points in d-dimensional space,
        fit an d-dimensional plane to the points.
        Return a point, p, on the plane (the point-cloud centroid),
        and the normal, n.
        """
        points = np.reshape(points, (np.shape(points)[0], -1))  # Collapse trialing dimensions
        assert points.shape[0] <= points.shape[1], "There are only {} points in {} dimensions.".format(points.shape[1],  points.shape[0])
        ctr = points.mean(axis=1)
        x = points - ctr[:, np.newaxis]
        M = np.dot(x, x.T)  # Could also use np.cov(x) here.
        return ctr, np.linalg.svd(M)[0][:, -1]

    def drawSamples(self):
        pass

File: cpp/src/SampleGrid.cpp
//
// Created by burban on 30.12.20.
//

#include "SampleGrid.h"
#include "constants.h"

Eigen::Vector3d  Sample::up = Eigen::Vector3d{0,0,0};
SampleGrid::SampleGrid() {
    this->upVector(2) = 1.0;
    this->floorPlane.n = this->upVector;


}

void SampleGrid::init() {

    this->samples.clear();
    for (int k = 0; k < CONSTANTS::FLOORDETECTION::SAMPLESY; k++) {
        std::vector<Sample> V;
        for (int l = 0; l < CONSTANTS::FLOORDETECTION::SAMPLESX; l++) {
            int i = l * (CONSTANTS::CAMERA::IMAGE_WIDTH - 1) / (CONSTANTS::FLOORDETECTION::SAMPLESX - 1);
            int j = CONSTANTS::CAMERA::IMAGE_HEIGHT - 1 - k * (CONSTANTS::CAMERA::IMAGE_HEIGHT - 1) / (
                    CONSTANTS::FLOORDETECTION::SAMPLESY - 1);
            Sample sample;
            sample.gridIdx = Eigen::Vector2i{l, k};

            sample.bufferIdx = i + j * CONSTANTS::CAMERA::IMAGE_WIDTH;
            V.emplace_back(sample);
        }
        samples.emplace_back(V);
    }
}

void SampleGrid::setUpVector(const Eigen::Vector3d &up) {
    this->upVector = up;
    this->upVector.normalize();
    Sample::up = upVector;
}

bool SampleGrid::isIn(const Eigen::Vector2i &gridIdx) const {
    return samples[gridIdx(1)][gridIdx(0)].inGrid;
}

void SampleGrid::update(const std::vector<Eigen::Vector3d> & points) {
    if (upVector.dot(floorPlane.n) > 0.5) {
        setUpVector(floorPlane.n);
    }

    for (int i = 0; i < samples.size(); i++) {
        for (int j = 0; j < samples[i].size(); j++) {
            samples[i][j].p = points.at(samples[i][j].bufferIdx);
            samples[i][j].inGrid = samples[i][j].p.any();
        }
    }

    // Compute all normals.
    Eigen::Vector3d normal;
    for (int i = 0; i < samples.size(); i++) {
        for (int j = 0; j < samples[i].size(); j++) {
            if (!samples[i][j].inGrid)
                continue;

            int upIdx = i + 1;
            if (i == samples.size() - 1 || !samples[upIdx][j].inGrid)
                upIdx = i;

            int downIdx = i - 1;
            if (i == 0 || !samples[downIdx][j].inGrid)
                downIdx = i;

            int rightIdx = j + 1;
            if (j == samples[i].size() - 1 || !samples[i][rightIdx].inGrid)
                rightIdx = j;

            int leftIdx = j - 1;
            if (j == 0 || !samples[i][leftIdx].inGrid)
                leftIdx = j;

            if (upIdx == downIdx || leftIdx == rightIdx)
                continue;

            Sample &up = samples[upIdx][j];
            Sample &down = samples[downIdx][j];
            Sample &right = samples[i][rightIdx];
            Sample &left = samples[i][leftIdx];
            normal =  -((up.p - down.p).cross((right.p - left.p)));
            normal.normalize();
            samples[i][j].n = normal;
        }
    }

}

Sample SampleGrid::findFloor() {
    prune();

    if (prunedSamples.size() < 2)
        return floorPlane;

    // Sort by z coordinate.
    sort(this->prunedSamples.begin(), this->prunedSamples.end(),
         [](const Sample &a, const Sample &b) -> bool {
             return a.p(2) < b.p(2);
         });


    // Reset things.
    planes.clear();
    planeAvg.clear();
    floorSegment.clear();
    floorPlane = prunedSamples[1]; // Always accept the first cluster.

    // Start a flood fill at every point in the sorted pruned set.
    for (int i = 2; i < prunedSamples.size() - 1; i++) {
        if (!isIn(prunedSamples[i].gridIdx))
            continue;

        planeCluster.clear();
        floodFill(prunedSamples[i].gridIdx);
        if (planeCluster.size() == 1)
            continue;


        Sample avg;
        avg.n(2) = 0;
        for (int j = 0; j < planeCluster.size(); j++)
            avg += planeCluster[j];
        avg /= planeCluster.size();

        planeAvg.push_back(avg);
        planes.push_back(planeCluster);


        if (floorPlane.distance(avg) < CONSTANTS::FLOORDETECTION::MERGE_THRESHOLD) {
            floorPlane.p = (floorPlane.p * floorSegment.size() + avg.p * planeCluster.size()) /
                           (floorSegment.size() + planeCluster.size());
            floorPlane.n = (floorPlane.n * floorSegment.size() + avg.n * planeCluster.size()) /
                           (floorSegment.size() + planeCluster.size());
            floorPlane.n.normalize();
            floorSegment.insert(floorSegment.end(), planeCluster.begin(), planeCluster.end());

        }

        // If that didn't work, a huge cluster can replace a very small one.
        if (floorSegment.size() * 20 < planeCluster.size()) {
            floorPlane = avg;
            floorSegment.clear();
            floorSegment.insert(floorSegment.end(), planeCluster.begin(), planeCluster.end());

        }
    }

    if (floorSegment.size() > 2) {
        const int cols = floorSegment.size();
        //const Eigen::Matrix< double , Eigen::Dynamic, Eigen::Dynamic > A(3, cols);
        //Eigen::Matrix<double,3,Eigen::Dynamic,Eigen::ColMajor> A;
        Eigen::MatrixX3d A;
        A.resize(cols,3);
        Eigen::VectorXd b;
        b.resize(cols);
        for (int i = 0; i < floorSegment.size(); i++) {
            A.row(i) = Eigen::Vector3d{floorSegment[i].p(0), floorSegment[i].p(1),1};
            b(i) = floorSegment[i].p(2);
        }

        Eigen::Vector3d solution =A.colPivHouseholderQr().solve(b);
        floorPlane.p(2) = floorPlane.p(0) * solution(0) + floorPlane.p(1) * solution(1) + solution(2);
        Eigen::Vector3d new_normal{-solution(0), -solution(1), 1};
        new_normal.normalize();
        floorPlane.n = new_normal;



    }
    return floorPlane;
}


void SampleGrid::floodFill(const Eigen::Vector2i &parentIdx) {
    Sample &parent = samples[parentIdx(1)][parentIdx(0)];
    if (!parent.inGrid)
        return;

    parent.inGrid = false;
    planeCluster.push_back(parent);


    if (parent.gridIdx(0) > 0) {
        Eigen::Vector2i childIdx = parent.gridIdx - Eigen::Vector2i(1, 0);
        Sample &child = samples[childIdx(1)][childIdx(0)];

        if (parent.distance(child) < CONSTANTS::FLOORDETECTION::FLOOD_THRESHOLD)
            floodFill(childIdx);
    }
    if (parent.gridIdx(0) < CONSTANTS::FLOORDETECTION::SAMPLESX - 1) {
        Eigen::Vector2i childIdx = parent.gridIdx + Eigen::Vector2i(1, 0);
        Sample &child = samples[childIdx(1)][childIdx(0)];
        if (parent.distance(child) < CONSTANTS::FLOORDETECTION::FLOOD_THRESHOLD)
            floodFill(childIdx);
    }
    if (parent.gridIdx(1) > 0) {
        Eigen::Vector2i childIdx = parent.gridIdx - Eigen::Vector2i(0, 1);
        Sample &child = samples[childIdx(1)][childIdx(0)];
        if (parent.distance(child) < CONSTANTS::FLOORDETECTION::FLOOD_THRESHOLD)
            floodFill(childIdx);
    }
    if (parent.gridIdx(1) < CONSTANTS::FLOORDETECTION::SAMPLESY - 1) {
        Eigen::Vector2i childIdx = parent.gridIdx + Eigen::Vector2i(0, 1);
        Sample &child = samples[childIdx(1)][childIdx(0)];

        if (parent.distance(child) < CONSTANTS::FLOORDETECTION::FLOOD_THRESHOLD)
            floodFill(childIdx);
    }
}

void SampleGrid::prune() {
    prunedSamples.clear();
    for (int i = 0; i < samples.size(); i++) {
        for (int j = 0; j < samples[i].size(); j++) {
            if (!samples[i][j].inGrid)
                continue;

            samples[i][j].angle = samples[i][j].n.dot(upVector); // A scalar product-based upright check.
            if (samples[i][j].angle > CONSTANTS::FLOORDETECTION::PRUNE_THRESHOLD) {
                prunedSamples.push_back(samples[i][j]);
            } else {
                samples[i][j].inGrid = false;
            }
        }
    }
}
